src/json/json_parser.c

```diff
@@ -0,0 +1,260 @@
+/*******************************************************************************
+*   (c) 2018, 2019 Zondax GmbH
+*
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+********************************************************************************/
+
+#include <jsmn.h>
+#include <zxmacros.h>
+#include <common/parser_common.h>
+#include "json_parser.h"
+
+#define EQUALS(_P, _Q, _LEN) (MEMCMP( (const void*) PIC(_P), (const void*) PIC(_Q), (_LEN))==0)
+
+parser_error_t json_parse(parsed_json_t *parsed_json, const char *buffer, uint16_t bufferLen) {
+    jsmn_parser parser;
+    jsmn_init(&parser);
+
+    MEMZERO(parsed_json, sizeof(parsed_json_t));
+    parsed_json->buffer = buffer;
+    parsed_json->bufferLen = bufferLen;
+
+    int32_t num_tokens = jsmn_parse(
+            &parser,
+            parsed_json->buffer,
+            parsed_json->bufferLen,
+            parsed_json->tokens,
+            MAX_NUMBER_OF_TOKENS);
+
+#ifdef APP_TESTING
+    char tmpBuffer[100];
+    snprintf(tmpBuffer, sizeof(tmpBuffer), "tokens: %d", num_tokens);
+    zemu_log(tmpBuffer);
+#endif
+
+    if (num_tokens < 0) {
+        switch (num_tokens) {
+            case JSMN_ERROR_NOMEM:
+                return parser_json_too_many_tokens;
+            case JSMN_ERROR_INVAL:
+                return parser_unexpected_characters;
+            case JSMN_ERROR_PART:
+                return parser_json_incomplete_json;
+            default:
+                return parser_json_unexpected_error;
+        }
+    }
+
+    parsed_json->numberOfTokens = 0;
+    parsed_json->isValid = 0;
```

Comment:
Should those 2 lines be moved at the beginning of the function to ensure proper initialization?